contracts/SurplusBeneficiary.sol

```diff
@@ -0,0 +1,131 @@
+    OwnableUpgradeable,
+    SurplusBeneficiaryStorageV1
+{
+    using AddressUpgradeable for address;
+    using SafeMathUpgradeable for uint256;
+
+    /// @dev this function is public for testing
+    // solhint-disable-next-line func-order
+    function initialize(
+        address tokenArg,
+        address feeDistributorArg,
+        address treasuryArg,
+        uint24 treasuryPercentageArg
+    ) public initializer {
+        __ReentrancyGuard_init();
+        __Ownable_init();
+
+        setToken(tokenArg);
+        setFeeDistributor(feeDistributorArg);
+        setTreasury(treasuryArg);
+        setTreasuryPercentage(treasuryPercentageArg);
+
+        IERC20Upgradeable(_token).approve(feeDistributorArg, uint256(-1));
+    }
+
+    //
+    // PUBLIC ONLY-ADMIN NON-VIEW
+    //
+
+    function setToken(address tokenArg) public onlyOwner {
+        // SB_TANC: token address is not contract
+        require(tokenArg.isContract(), "SB_TANC");
+        emit TokenChanged(_token, tokenArg);
+        _token = tokenArg;
+    }
+
+    function setFeeDistributor(address feeDistributorArg) public onlyOwner {
+        // SB_FDNC: feeDistributor address is not contract
+        require(feeDistributorArg.isContract(), "SB_FDNC");
+        emit FeeDistributorChanged(_feeDistributor, feeDistributorArg);
+        _feeDistributor = feeDistributorArg;
+    }
+
+    function setTreasury(address treasuryArg) public onlyOwner {
+        // SB_TNC: treasury address is not contract
+        require(treasuryArg.isContract(), "SB_TNC");
+        emit TreasuryChanged(_treasury, treasuryArg);
+        _treasury = treasuryArg;
+    }
+
+    function setTreasuryPercentage(uint24 treasuryPercentageArg) public onlyOwner {
+        // SB_TPZ: treasury percentage is equal to zero
+        require(treasuryPercentageArg != 0, "SB_TPZ");
+        emit TreasuryPercentageChanged(_treasuryPercentage, treasuryPercentageArg);
+        _treasuryPercentage = treasuryPercentageArg;
+    }
+
+    //
+    // EXTERNAL NON-VIEW
+    //
+
+    /// @inheritdoc ISurplusBeneficiary
+    function dispatch() external override nonReentrant returns (uint256) {
+        address token = _token;
+
+        uint256 tokenAmount = IERC20Upgradeable(token).balanceOf(address(this));
+
+        // SB_TAZ: token amount is zero
+        require(tokenAmount != 0, "SB_TAZ");
+
+        uint256 tokenAmountToTreasury = FullMath.mulDiv(tokenAmount, _treasuryPercentage, 1e6);
+
+        // transfer to treasury first, because FeeDistributor.burn() will transfer all balance from SurplusBeneficiary
+        SafeERC20Upgradeable.safeTransfer(IERC20Upgradeable(token), _treasury, tokenAmountToTreasury);
+        IFeeDistributor(_feeDistributor).burn(token);
+
+        uint256 balanceAfter = IERC20Upgradeable(token).balanceOf(address(this));
+
+        // SB_BNZ: balance is not zero
+        require(balanceAfter == 0, "SB_BNZ");
+
+        emit Dispatch(tokenAmountToTreasury, tokenAmount.sub(tokenAmountToTreasury));
```

Comment:
i don't know what's easier for subgraph, but do we also need to record the amount when we can query the amount from ERC20 transfer?
maybe emit target address is more useful, if the target address may be changed by admin

Reply:
a dedicated event is more self-descriptive and more future proof (that we don't need to remember the assumptions we made to the ERC20 transfers in terms of analytics)